dags/process_elife_zip_dag.py

```diff
@@ -113,47 +89,90 @@ def extract_archived_files_to_bucket(**context):
             raise FileNotFoundError(error_message)
 
     # pass article cloud bucket key to next task
     return '%s/%s' % (folder_name, article_name)
 
 
-def wrap_article_in_libero_xml_and_send_to_service(**context):
-    # get xml path passed from previous task
-    previous_task = get_previous_task_name(**context)
-    xml_path = context['task_instance'].xcom_pull(task_ids=previous_task)
-    logger.info('XML PATH PASSED FROM PREVIOUS TASK= %s', xml_path)
-    message = 'path to xml document was not passed from task %s' % previous_task
-    assert xml_path is not None, message
+def convert_tiff_images_in_expanded_bucket_to_jpeg_images(**context):
+    zip_file_name = get_file_name_passed_to_dag_run_conf_file(**context)
+    for key in list_bucket_keys_iter(Bucket=DESTINATION_BUCKET, Prefix=zip_file_name.rstrip('.zip')):
+        if key.endswith('.tif'):
+            with TemporaryFile(dir=TEMP_DIRECTORY) as temp_tiff_file:
+                s3 = get_aws_connection('s3')
+                s3.download_fileobj(
+                    Bucket=DESTINATION_BUCKET,
+                    Key=key,
+                    Fileobj=temp_tiff_file
+                )
+                image = Image.open(temp_tiff_file)
+                output = image.convert('RGB')
+                temp_jpeg = BytesIO()
+                output.save(temp_jpeg, 'JPEG')
+                key = re.sub(r'\.\w+$', '.jpg', key)
+                s3.put_object(
+                    Bucket=DESTINATION_BUCKET,
+                    Key=key,
+                    Body=temp_jpeg.getvalue()
+                )
+                logger.info('%s uploaded to %s',key, DESTINATION_BUCKET)
+
 
+def update_tiff_references_to_jpeg_in_article(**context):
+    zip_file_name = get_file_name_passed_to_dag_run_conf_file(**context)
+    article_name = get_expected_elife_article_name(zip_file_name)
+    folder_name = zip_file_name.rstrip('.zip')
+    s3_key = '%s/%s' % (folder_name, article_name)
     s3 = get_aws_connection('s3')
-    response = s3.get_object(Bucket=DESTINATION_BUCKET, Key=xml_path)
+    response = s3.get_object(Bucket=DESTINATION_BUCKET, Key=s3_key)
+    article_bytes = BytesIO(response['Body'].read())
+    article_xml = etree.parse(article_bytes)
+
+    # save a copy of the original xml document before making modifications
+    key = s3_key.replace('.xml', '-original.xml')
+    s3.put_object(Bucket=DESTINATION_BUCKET, Key=key, Body=article_bytes.getvalue())
+    logger.info('%s uploaded to %s/%s', key, DESTINATION_BUCKET, folder_name)
+
+    for element in article_xml.xpath(
+            '//*[@mimetype="image" and @mime-subtype="tiff"]'):
+        href = '{http://www.w3.org/1999/xlink}href'
+        element.attrib[href] = re.sub(r'\.\w+$', '.jpg', element.attrib[href])
+        element.attrib['mime-subtype'] = 'jpeg'
+
+    # upload modified document
+    s3.put_object(
+        Bucket=DESTINATION_BUCKET,
+        Key=s3_key,
```

Comment:
S3 in general has [eventual consistency on updates of keys](https://docs.aws.amazon.com/AmazonS3/latest/dev/Introduction.html#ConsistencyModel); so renaming would be fine (a new key is created), the overwrite may not (clients continuing to see the older content).
The way to treat keys to avoid these problems is to think of them as immutable, hence `s3-key` should be a new value like `elife-666-converted-images.xml` or any other suffix. This requires, however, to pass the new file name to the next task so that it does not assume what the file name is. Will be interesting to see how Airflow allows multiple tasks (the original extraction and this) to store a value for a downstream task.